### PR TITLE
Move package metadata and tool configuration to `pyproject.toml`

### LINKED ISSUE #1126: Update and use `bumpversion` or otherwise automate syncing versions
At the moment we have version 1.0.0 listed in some places and version 0.1.0 in others

https://github.com/UCL/TLOmodel/blob/46100d2eff21fe4c0e7ea253d8aae6f229884a6f/setup.py#L21
https://github.com/UCL/TLOmodel/blob/46100d2eff21fe4c0e7ea253d8aae6f229884a6f/CITATION.cff#L6
https://github.com/UCL/TLOmodel/blob/46100d2eff21fe4c0e7ea253d8aae6f229884a6f/docs/conf.py#L42
https://github.com/UCL/TLOmodel/blob/46100d2eff21fe4c0e7ea253d8aae6f229884a6f/.bumpversion.cfg#L2
https://github.com/UCL/TLOmodel/blob/46100d2eff21fe4c0e7ea253d8aae6f229884a6f/src/tlo/__init__.py#L15

`bumpversion` is I think designed to specifically to address this problem by having one place the version is set which is then propagated to other files, however we haven't been using it recently  as far as I know (for example it's not included in the `dev` requirements) and [the version we are currently using is no longer actively developed](https://github.com/peritus/bumpversion) though [there is a fork which has more recent activity](https://github.com/c4urself/bump2version/).

If we want to continue with using `bumpversion` / `bump2version` we should 
  * Add the version tag in the `CITATION.cff` as an additional rule to the configuration file
  * Either have the `bumpversion` in dry run mode included as part of the checks (to see if any files are changed and fail if so) or possibly as a precommit hook

Alternatively we could look at other approaches for automating versioning like [setuptools-scm](https://pypi.org/project/setuptools-scm/)

### LINKED ISSUE #1156: Add `pylint` to checks to pick up on some statically identifiable errors
Following on from https://github.com/UCL/TLOmodel/pull/1155#issuecomment-1755739896, running `pylint --disable=R,C,W src/tlo` on current `master` outputs

```
************* Module src.tlo.core
src/tlo/core.py:243:34: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
************* Module src.tlo.cli
src/tlo/cli.py:182:21: E1125: Missing mandatory keyword argument 'type' in constructor call (missing-kwoa)
src/tlo/cli.py:779:4: E1120: No value for argument 'ctx' in function call (no-value-for-parameter)
src/tlo/cli.py:779:4: E1120: No value for argument 'config_file' in function call (no-value-for-parameter)
src/tlo/cli.py:779:4: E1120: No value for argument 'verbose' in function call (no-value-for-parameter)
************* Module src.tlo.util
src/tlo/util.py:64:81: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
src/tlo/util.py:94:45: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
src/tlo/util.py:206:58: E1130: bad operand type for unary ~: int64 (invalid-unary-operand-type)
************* Module src.tlo.scenario
src/tlo/scenario.py:263:28: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
************* Module src.tlo.lm
src/tlo/lm.py:385:22: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
************* Module src.tlo.simulation
src/tlo/simulation.py:82:19: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
src/tlo/simulation.py:164:25: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
************* Module src.tlo.methods.healthsystem
src/tlo/methods/healthsystem.py:803:33: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
src/tlo/methods/healthsystem.py:804:26: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
src/tlo/methods/healthsystem.py:805:30: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
************* Module src.tlo.methods.contraception
src/tlo/methods/contraception.py:281:20: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
src/tlo/methods/contraception.py:591:27: E1136: Value 'p_pregnancy_with_contraception_per_month' is unsubscriptable (unsubscriptable-object)
************* Module src.tlo.methods.healthburden
src/tlo/methods/healthburden.py:328:14: E1136: Value 'self.decompose_yll_by_age_and_time(start_date=date_of_death, end_date=min(self.sim.end_date, date_of_birth + pd.DateOffset(years=self.parameters['Age_Limit_For_YLL'])), date_of_birth=date_of_birth)' is unsubscriptable (unsubscriptable-object)
************* Module src.tlo.methods.scenario_switcher
src/tlo/methods/scenario_switcher.py:61:20: E0702: Raising NoneType while only classes or instances are allowed (raising-bad-type)
************* Module src.tlo.methods.hiv
src/tlo/methods/hiv.py:3172:4: E1137: 'x' does not support item assignment (unsupported-assignment-operation)
src/tlo/methods/hiv.py:3172:35: E1136: Value 'x' is unsubscriptable (unsubscriptable-object)
************* Module src.tlo.methods.chronicsyndrome
src/tlo/methods/chronicsyndrome.py:290:13: E1101: Module 'numpy.random' has no 'RandomState' member (no-member)
************* Module src.tlo.logging.helpers
src/tlo/logging/helpers.py:29:52: E1101: Instance of 'RootLogger' has no 'loggerDict' member (no-member)
************* Module src.tlo.analysis.utils
src/tlo/analysis/utils.py:99:13: E0110: Abstract class 'ExcelWriter' with abstract methods instantiated (abstract-class-instantiated)
src/tlo/analysis/utils.py:109:4: E1101: Instance of 'ExcelWriter' has no 'save' member; maybe '_save'? (no-member)
************* Module src.tlo.test.random_birth
src/tlo/test/random_birth.py:88:52: E1101: Instance of 'RandomBirth' has no 'pregnancy_probability' member (no-member)
```

which picks up the issue in #1154. There are also some what appear to be false positives, but there are sufficiently few that we could probably go through and add `# noqa` exceptions for these. There are also some which looks like they may be actual errors.

### PR DESCRIPTION
**Pull request by matt-graham**
Fixes #1126 and fixes #1156

Moves from `setup.py` / `setup.cfg` based approach of specifying package metadata and tool configuration to more modern `pyproject.toml` based approach.

This allows getting rid of various separate configuration / metadata files (`.coveragerc`, `requirements/*.in`, `setup.cfg`, `setup.py`) in favour having all configuration centralized in `pyproject.toml`.

This PR also removes the previous `.bumpversion.cfg` file in favour of moving to [`setuptools-scm` ](https://github.com/pypa/setuptools_scm) to automatically set package version based on Git metadata / tags ([default versioning scheme is described here](https://setuptools-scm.readthedocs.io/en/latest/usage/#default-versioning-scheme)).

The `tox` `check` environment is updated to use [Ruff](https://docs.astral.sh/ruff/) in place of `flake8` as Ruff is a lot quicker and supports `pyproject.toml` which [`flake8` currently doesn't](https://github.com/PyCQA/flake8/issues/234). For now to keep the changes minimal here, I've added a series of rules to ignore that cause the Ruff check to fail. In a separate PR we probably want to either move to adding specific `# noqa: ...` overrides at the relevant places if we think what we have in the code is correct or fix the violations.

I've also removed the previous `requirements/Makefile` in favour of adding a new `requirements` `tox` environment that updates the requirements files using `pip-compile` (with input requirements now sourced from dependencies specified in `pyproject.toml`, including optional `dev` dependencies). I've also regenerated the requirements files so that the commented sources / generation command align with new structure but as far as I can see from a quick check this hasn't led to any changes in resolved packages. (EDIT: Beyond a few changes from adding `ruff` to development dependencies).

EDIT: Now also adds `pylint` to checks, as for `ruff` disabling rules that currently cause checks to fail with intention of fixing actual violations in separate PRs and/or adding specific `# noqa: ...` overrides.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,119 @@
+[build-system]
+build-backend = "setuptools.build_meta"
+requires = ["setuptools>=62", "setuptools_scm[toml]>=8.0"]
+
+[project]
+name = "tlo"
+authors = [
+    {name = "Thanzi La Onse Model Development Team", email="a.tamuri@ucl.ac.uk"},
```
Comment (matt-graham) [inline pyproject.toml:8]: @tamuri I copied this across from previous `setup.py` but we could either remove `email` field if your prefer or use `arc.collaborations@ucl.ac.uk` for example?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+]
+
+[tool.isort]
+force_single_line = false
+multi_line_output = 3
+include_trailing_comma = true
+line_length = 100
+known_first_party = "tlo"
+default_section = "THIRDPARTY"
+skip = "migrations"
```
[the remark this answers is not in the document]
Comment (tamuri) [reply pyproject.toml:89]: yes, that can be removed (both mentions)

File: pyproject.toml (hunk 1 above)
Comment (tamuri) [reply pyproject.toml:8]: We can remove email field, thanks.

**Comment by tbhallett** [comment]
Thanks for this one @matt-graham. Just noting that this one is the critical one for the first release with the paper submission

**Review by tamuri — APPROVED** [review]

**Pull request merged**